src/auth/index.js

```diff
@@ -36,14 +61,29 @@ const env = ({ authBucket, generateAuthBucket }, next) => (req, res) => {
     auth: oauth,
   }, {}, (err, response) => {
     if (response && response.permissions && response.permissions.includes(permission)) {
-      next(req, res);
+      resolve();
     } else {
-      res.status(403).json({
-        code: 403,
-        message: `The request is forbidden. ${bucket} did not allow access`,
-      });
+      reject({ code: 403, message: `The request is forbidden. ${bucket} did not allow access` });
     }
   });
+});
+
+const env = ({ authBucket, generateAuthBucket }, next) => (req, res) => {
+  let oauth = null;
+  return getAccessToken(req.get('Authorization'))
+    .then(retrieveUserInformation)
+    .then((info) => {
+      if (!req.body) req.body = {};
+      req.body = Object.assign(req.body, { gcframe: info });
+    })
+    .then(() => { oauth = generateGoogleOAuth(req.body.gcframe.accessToken); })
+    .then(() => deriveGoogleStorageBucket(authBucket, generateAuthBucket, req))
+    .then((bucket) => checkUserPermissionForStorageBucket(bucket, oauth))
+    .then(() => next(req, res))
+    .catch((err) => {
```

Comment:
I'm concerned we've made the catch/error flow here incorrect (or more difficult to understand in any case).

If there is a error in the `next` handler, we will end up in the catch here, no? Then `err.code` is undefined I guess.

I noticed in the dupe tags PR also, that putting a single `catch` at the end of the Promise chain lead to problems. It's like putting a big try around a somewhat large code block and then trying to handle every possible error in one place.

Reply:
I think we might need to have a sketchup for this... personally I think having a error hoover is a good thing - we can always define a filter function to sit within the catch to do what we want in terms of enforcing correct err responses to a client (i.e 403 unauthorized) and then have 500 errors for anything that we're not expecting (maybe an error caused by next func because the person implementing the middleware hasn't defended well). This approach also allows us to basically never crash, just log and accept the error has happened.